### PR TITLE
Serialize Trampoline payloads in outbound onions.

### PR DESCRIPTION
**Pull request by arik-so**
This PR alters the outbound onion payload to optionally accommodate a Trampoline packet.

Additionally, we add the Trampoline support feature flag.

### CODE REVIEW

**Review by shaavan — COMMENTED** [review]
The first draft of the trampoline support looks good.

Referring to Spec PR [requirements](https://github.com/lightning/bolts/blob/trampoline-onion/04-onion-routing.md#requirements-2):

1. Correctly implements the `trampoline_packet` field to allow receiving of trampoline onion packets
2. Ensure the same construction of `trampoline_packets` as the `OnionPacket`.

I have a few questions I would love to discuss. Looking forward to your insights.

File: lightning/src/ln/msgs.rs [hunk 1]
```diff
@@ -2276,6 +2278,7 @@ impl Writeable for OutboundOnionPayload {
 					(2, HighZeroBytesDroppedBigSize(*amt_msat), required),
 					(4, HighZeroBytesDroppedBigSize(*outgoing_cltv_value), required),
 					(8, payment_data, option),
+					(12, trampoline_packet, option),
```
[the remark this answers is not in the document]
Comment (t-bast) [reply lightning/src/ln/msgs.rs:2281]: Yes, all the numbers in the current spec PR are obsolete, they are from many years ago and some of the TLV tags have already been allocated to other features. I was waiting for another implementation to seriously look at Trampoline before re-working the numbers. For now, I think you should use placeholder numbers here (large ones to avoid clashes), that will be replaced by the final numbers once we are ready for cross-compat tests!

File: lightning/src/ln/features.rs [hunk 2]
```diff
@@ -415,6 +417,9 @@ mod sealed {
 	define_feature!(55, Keysend, [NodeContext],
 		"Feature flags for keysend payments.", set_keysend_optional, set_keysend_required,
 		supports_keysend, requires_keysend);
+	define_feature!(57, Trampoline, [NodeContext],
```
Comment (shaavan) [inline lightning/src/ln/features.rs:420]: Trampoline is introduced as Feature of NodeContext, but according to [BOLT#9](https://github.com/lightning/bolts/blob/086a8af2943a721b99e818c2c6d880efc5376688/09-features.md) it should of _IN9_ context, that is InitContext, NodeContext, and Bolt11InvoiceContext. Is the rest of the context planned to be introduced later?

File: lightning/src/ln/msgs.rs [hunk 3]
```diff
@@ -2277,13 +2319,17 @@ impl Writeable for OutboundOnionPayload {
...
 				ref payment_data, ref payment_metadata, ref keysend_preimage, amt_msat,
-				outgoing_cltv_value, ref custom_tlvs,
+				outgoing_cltv_value, ref trampoline_packet, ref custom_tlvs
 			} => {
 				// We need to update [`ln::outbound_payment::RecipientOnionFields::with_custom_tlvs`]
 				// to reject any reserved types in the experimental range if new ones are ever
 				// standardized.
 				let keysend_tlv = keysend_preimage.map(|preimage| (5482373484, preimage.encode()));
-				let mut custom_tlvs: Vec<&(u64, Vec<u8>)> = custom_tlvs.iter().chain(keysend_tlv.iter()).collect();
+				let trampoline_tlv = trampoline_packet.as_ref().map(|trampoline| (66100, trampoline.encode()));
```
Comment (shaavan) [inline lightning/src/ln/msgs.rs:2328]: Out of curiosity, I was able to find a reference for why keysend why keyed by `5482373484` in [blip-0003](https://github.com/lightning/blips/blob/master/blip-0003.md), but I was able to find no such reference for `66100`, for the trampoline. Is it an arbitrary choice for now, or is there some reasoning behind it?

File: lightning/src/ln/msgs.rs [hunk 4]
```diff
@@ -1787,6 +1789,29 @@ impl fmt::Debug for OnionPacket {
 	}
 }
 
+/// BOLT 4 onion packet including hop data for the next peer.
+#[derive(Clone, Hash, PartialEq, Eq)]
+pub struct VariableLengthOnionPacket {
```
Comment (shaavan) [inline lightning/src/ln/msgs.rs:1794]: As per [BOLT#9](https://github.com/lightning/bolts/blob/trampoline-onion/04-onion-routing.md#trampoline-payments), the Trampoline is correctly introduced as a `VariableLengthOnionPacket`

```
The trampoline_onion_packet has a variable size to allow...
```

The struct looks good and is correctly implemented. And, I think it can use a little nudge on the documentation side so that its purpose is explained in more detail and it can be easily distinguishable on a documentation level from struct `OnionPacket`.

**Review by shaavan — COMMENTED** [review]
First pass.
So far, the PR looks good.
Thanks for the awesome work!!

Writing down my thoughts here and discussing some questions I have:

File: lightning/src/ln/msgs.rs (hunk 3 above)
Comment (shaavan) [reply lightning/src/ln/msgs.rs:2328]: Thanks a lot for the reference!

I think it will be a good idea to write a comment about it, explaining the reasoning behind it and the necessary caution one should keep.
Eclair's comment seems like a pretty solid one to take a reference from.

> // Types below aren't specified - use cautiously when deploying (be careful with backwards-compatibility).

File: lightning/src/ln/features.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (shaavan) [reply lightning/src/ln/features.rs:420]: Thanks a lot for the reference!

File: lightning/src/ln/msgs.rs [hunk 5]
```diff
@@ -1789,6 +1789,29 @@ impl fmt::Debug for OnionPacket {
...
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:1794]: Can we consolidate this with `onion_message::packet::Packet`?

**Review by vladimirfomene — COMMENTED** [review]
Like Matt, I'm also wondering what is the reason for using the onion message packet.

File: lightning/src/ln/msgs.rs [hunk 6]
```diff
@@ -1698,6 +1698,13 @@ mod fuzzy_internal_msgs {
 			amt_to_forward: u64,
 			outgoing_cltv_value: u32,
 		},
+		#[allow(dead_code)] // TODO (arik): drop in follow-up PR
+		TrampolineEntry {
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:1702]: IMO the word "Entry" is superfluous. Unless there are multiple trampoline outbound payload variants?

File: lightning/src/ln/msgs.rs [hunk 7]
```diff
@@ -1698,6 +1698,13 @@ mod fuzzy_internal_msgs {
...
+			amt_to_forward: u64,
+			outgoing_cltv_value: u32,
+			payment_data: Option<FinalOnionHopData>,
+			trampoline_packet: crate::onion_message::packet::Packet
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:1706]: I thought we didn't like the `onion_message::` module prefix here?

File: lightning/src/ln/msgs.rs [hunk 8]
```diff
@@ -2486,7 +2504,9 @@ impl Writeable for OutboundOnionPayload {
 				// to reject any reserved types in the experimental range if new ones are ever
 				// standardized.
 				let keysend_tlv = keysend_preimage.map(|preimage| (5482373484, preimage.encode()));
-				let mut custom_tlvs: Vec<&(u64, Vec<u8>)> = custom_tlvs.iter().chain(keysend_tlv.iter()).collect();
+				let mut custom_tlvs: Vec<&(u64, Vec<u8>)> = custom_tlvs.iter()
+					.chain(keysend_tlv.iter())
+					.collect();
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:2509]: Looks like unnecessary diff

File: lightning/src/ln/msgs.rs [hunk 9]
```diff
@@ -4338,6 +4359,64 @@ mod tests {
 		} else { panic!(); }
 	}
 
+	#[test]
+	fn encoding_final_onion_hop_data_with_trampoline_packet() {
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:4363]: Are these tests from the spec? Would be good to have some high level comments for them

File: lightning/src/ln/features.rs [hunk 10]
```diff
@@ -66,6 +66,8 @@
 //!      for more info).
 //! - `Keysend` - send funds to a node without an invoice
 //!     (see the [`Keysend` feature assignment proposal](https://github.com/lightning/bolts/issues/605#issuecomment-606679798) for more information).
+//! - `Trampoline` - send funds to a node without without knowing a direct path using nested onions
```
Comment (valentinewallace) [inline lightning/src/ln/features.rs:69]: Does the feature bit indicate support for forwarding as a trampoline node? This sounds like it's for sending.

File: lightning/src/ln/msgs.rs [hunk 11]
```diff
@@ -3046,7 +3066,7 @@ mod tests {
 	use crate::ln::msgs::{self, FinalOnionHopData, OnionErrorPacket, CommonOpenChannelFields, CommonAcceptChannelFields};
 	use crate::ln::msgs::SocketAddress;
 	use crate::routing::gossip::{NodeAlias, NodeId};
-	use crate::util::ser::{Writeable, Readable, ReadableArgs, Hostname, TransactionU16LenLimited};
+	use crate::util::ser::{Writeable, Readable, ReadableArgs, Hostname, TransactionU16LenLimited, BigSize};
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:3069]: super nit: alphabetize imports

**Review by valentinewallace — COMMENTED** [review]
Commit history doesn't make a ton of sense to me :( seems like there should be 1 commit for the feature bit, 1 for the payload variant and maybe another separate one for the tests?

File: lightning/src/ln/msgs.rs [hunk 12]
```diff
@@ -1698,6 +1698,13 @@ mod fuzzy_internal_msgs {
 			amt_to_forward: u64,
 			outgoing_cltv_value: u32,
 		},
+		#[allow(unused)]
+		TrampolineEntry {
+			amt_to_forward: u64,
+			outgoing_cltv_value: u32,
+			payment_data: Option<FinalOnionHopData>,
```
[the remark this answers is not in the document]
Comment (t-bast) [reply lightning/src/ln/msgs.rs]: All of those fields work exactly the same as for normal payments, as the behavior of an intermediate trampoline node, 
when processing HTLCs that contain a trampoline onion is to:

- first act as if it were a payment recipient, aggregating incoming payment parts
- then forward the "received" payment to a relayer component instead of treating it to a payment to us

Rusty also had a question about the `payment_secret` in the outer onion, as that part is a bit subtle: https://github.com/lightning/bolts/pull/836#discussion_r566557686.

In eclair we use the following trait hierarchy: https://github.com/ACINQ/eclair/blob/c8184b3e43a603454d672aed61b2ae461a80e1a8/eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/PaymentOnion.scala#L197. Once decrypted, we combine the outer and inner onion information received by an intermediate trampoline node as `NodeRelayPayload`, to show that this data is used for relay.

> @t-bast, do you figure it'd be worthwhile expanding a bit in the https://github.com/lightning/bolts/pull/836 on how the roles of the fields in the outer onion are altered in a Trampoline context?

I thought that was already detailed well enough in the requirements section of the specification, if there is a specific point you'd like to be more detailed can you put a comment on the spec?

> As well as any arithmetic considerations for routing fees for a path whose length isn't known a priori?

That's something I described in the "trampoline fees" section of this spec PR: https://github.com/lightning/bolts/pull/829
Let me know directly on that PR if the section is unclear or if you disagree with some of it. In eclair we previously used a very basic trial-and-error approach (start with a low fee, increase it on errors), and recently moved to a different model where the sender always uses 4% for the trampoline fees, and the trampoline node tries to deal with that (and could in the future do it at a loss in some payments as long as other payments compensate that loss).

File: lightning/src/ln/msgs.rs [hunk 13]
```diff
@@ -1698,6 +1698,13 @@ mod fuzzy_internal_msgs {
 			amt_to_forward: u64,
 			outgoing_cltv_value: u32,
 		},
+		#[allow(unused)]
+		TrampolineEntry {
+			amt_to_forward: u64,
+			outgoing_cltv_value: u32,
+			multipath_trampoline_data: Option<FinalOnionHopData>,
+			trampoline_packet: crate::onion_message::packet::Packet,
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:1706]: Did we figure out that this struct is reusable in the trampoline case? If so we should probably move it to a different module to avoid the `onion_message` module prefix.

File: lightning/src/ln/msgs.rs (hunk 6 above)
Comment (valentinewallace) [reply lightning/src/ln/msgs.rs:1702]: Ok, I see. I misunderstood "entry" to mean it's like, an entry into the vec of payloads. Maybe `Entrypoint` or `Initial` would be clearer? Could also add a comment.

**Review by valentinewallace — COMMENTED** [review]
Seems legit. Did we need to add a comment around the `allow(unused)` thing? Otherwise, I don't see any reason to hold this up.

File: lightning/src/ln/msgs.rs [hunk 14]
```diff
@@ -2478,6 +2485,17 @@ impl Writeable for OutboundOnionPayload {
 					(6, short_channel_id, required)
 				});
 			},
+			Self::TrampolineEntrypoint {
+				amt_to_forward, outgoing_cltv_value, multipath_trampoline_data: ref payment_data,
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:2489]: nit:
```suggestion
				amt_to_forward, outgoing_cltv_value, ref multipath_trampoline_data,
```
This is my one complaint about the auto-rename feature, that it isn't smart about this

**Review by valentinewallace — APPROVED** [review]
I'm fine with moving the `onion_message::Packet` into a different module later. Let's land her!

File: lightning/src/ln/msgs.rs [hunk 15]
```diff
@@ -1765,6 +1772,52 @@ impl fmt::Debug for OnionPacket {
...
+	/// Bolt 04 version number
+	pub version: u8,
+	/// A random sepc256k1 point, used to build the ECDH shared secret to decrypt hop_data
+	pub public_key: PublicKey,
+	/// Encrypted payload for the next hop
+	//
+	// Unlike the onion packets used for payments, Trampoline onion packets have to be shorter than
+	// 1300 bytes. The expected default is 650 bytes.
+	// TODO: if 650 ends up being the most common size, optimize this to be:
+	// enum { ThirteenHundred([u8; 650]), VarLen(Vec<u8>) }
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:1787]: nit: s/ThirteenHundred/SixFifty, I guess

**Review by valentinewallace — APPROVED** [review]
